src/engine.ml

```diff
@@ -1042,20 +1033,39 @@ let out ?add_timestamp (ctx : keys) hmac_algorithm compress rng data =
   (* - compression only if configured (0xfa for uncompressed)
      the ~add_timestamp argument is only used in static key mode
   *)
-  let replay_id =
-    let buf = Cstruct.create Packet.id_len in
-    Cstruct.BE.set_uint32 buf 0 ctx.my_replay_id;
-    buf
+  let set_replay_id dest off =
+    Cstruct.BE.set_uint32 dest off ctx.my_replay_id;
   in
-  let compression =
-    if compress then (
-      let buf = Cstruct.create 1 in
-      (* 0xFA is "no compression" *)
-      Cstruct.set_uint8 buf 0 0xfa;
-      buf)
-    else Cstruct.empty
+  let aead (type key)
+      (authenticate_encrypt_tag :
+         key: key -> nonce: Cstruct.t -> ?adata: Cstruct.t -> Cstruct.t -> Cstruct.t * Cstruct.t)
+      (my_key : key)
+      my_implicit_iv
+      =
+      let nonce, replay_id =
+        let b = Cstruct.create_unsafe (Packet.id_len + Cstruct.length my_implicit_iv) in
+        Cstruct.BE.set_uint32 b 0 ctx.my_replay_id;
+        Cstruct.blit my_implicit_iv 0 b 4 (Cstruct.length my_implicit_iv);
+        (* We reuse the replay id part of the nonce to avoid another allocation *)
+        b, Cstruct.sub b 0 4
+      in
+      let data =
+        let b = Cstruct.create (Bool.to_int compress + Cstruct.length data) in
+        (* 0xFA is "no compression" *)
+        Cstruct.memset (Cstruct.sub b 0 (Bool.to_int compress)) 0xfa;
+        Cstruct.blit data 0 b (Bool.to_int compress) (Cstruct.length data);
+        b
+      in
+      let enc, tag =
+        authenticate_encrypt_tag
+          ~key:my_key ~nonce ~adata:replay_id data
+      in
+      let res = Cstruct.create (4 + Cstruct.length tag + Cstruct.length enc) in
+      set_replay_id res 0;
+      Cstruct.blit tag 0 res 4 (Cstruct.length tag);
+      Cstruct.blit enc 0 res (4 + Cstruct.length tag) (Cstruct.length enc);
+      res
```

Comment:
```suggestion
      Cstruct.concat [ replay_id ; tag ; enc ]
```

Reply:
Ah yes, I initially removed `replay_id` but now we have it again (throug reuse). Thanks!